content/posts/pandastutor.md

```diff
@@ -0,0 +1,323 @@
+fortunately Pyodide comes with pandas
+[built-in](https://pyodide.org/en/stable/usage/packages-in-pyodide.html)!
+
+
+### Step 2: create a new Pyodide entry point
+
+The only modification we had to make to the Pandas Tutor codebase was
+adding a new `pyodide_main.py` file as the entry point for Pyodide:
+
+```py
+# simplified contents of pandas_tutor/pyodide_main.py
+from .main import run_user_code
+import js # to interface with pyodide
+def run_code_from_js():
+    return run_user_code(js.globalUserCode)
+```
+
+This entry point uses the `js` module to access JavaScript global
+variables. On the JavaScript end of our web frontend code, we created a
+corresponding function that takes the user's code to run (`userCode`)
+and calls the `run_code_from_js()` Python function:
+
+```js
+function runPyodidePandastutor(userCode) {
+  window.globalUserCode = userCode; // pass function argument as a global variable
+  const result = pyodide.runPython('pandas_tutor.pyodide_main.run_code_from_js()');
+  window.globalUserCode = undefined; // null this out afterward to prevent leakage
+  return result;
+}
+```
+
+We couldn't figure out a more elegant way to pass arguments from
+JavaScript to Python functions, so we used a `globalUserCode` variable
+as a hack that seems to work for now.
+
+
+### Step 3: import the Pandas Tutor module in JavaScript
+
+Our final setup step was adding a line of code to our frontend
+JavaScript to import the Pandas Tutor module:
+
+```js
+pyodide.runPython('import pandas_tutor.pyodide_main');
+```
+
+This is a synchronous call that imports `pandas_tutor` and all
+dependencies when the web app first loads. This line takes 3-5 seconds
+to finish since dependencies (including pandas) must be loaded.
+Unfortunately, browser caching can't speed things up here since even if
+all the files are cached locally, Pyodide still needs to run the import
+and initialization code at page load time.
+
+One feature addition that might speed up this process is if Pyodide
+could somehow *serialize a binary memory snapshot* taken right after
+everything gets initialized. Then the user's browser can directly load
+that snapshot on the first page visit and cache it for future visits.
+(But maybe that snapshot would be too big to download efficiently, so
+nevermind!)
+
+
+### Bonus: other fine-tuning and ongoing work
+
+- [Graceful
+  degradation](https://developer.mozilla.org/en-US/docs/Glossary/Graceful_degradation):
+  To continue supporting older browsers, our code detects if there are
+  any problems loading the Pyodide version of Pandas Tutor and, if so,
+  automatically reverts back to the original server-side version. Also,
+  [loadPyodide](https://pyodide.org/en/stable/usage/api/js-api.html#globalThis.loadPyodide)
+  crashes on some older mobile devices (presumably with [limited
+  memory](https://github.com/pyodide/pyodide/issues/533)?) and we can't
+  seem to catch the exception to recover; so we disable it by default on
```

Comment:
If you have an example of that, it would be great to open an issue. But yeah, that's unrelated to this post.